src/vm/analysis/type_checker/tests/mod.rs

```diff
@@ -1699,3 +1705,96 @@ fn test_set_entry_unbound_variables() {
+    let good = [
+        "(define-private (get-len (x (string-ascii 1)) (acc uint)) (+ acc u1))
+        (fold get-len \"blockstack\" u0)",
+        "(define-private (slice (x (string-ascii 1)) (acc (tuple (limit uint) (cursor uint) (data (string-ascii 10)))))
+            (if (< (get cursor acc) (get limit acc))
+                (let ((data (default-to (get data acc) (as-max-len? (concat (get data acc) x) u10))))
+                    (tuple (limit (get limit acc)) (cursor (+ u1 (get cursor acc))) (data data)))
+                acc))
+        (fold slice \"blockstack\" (tuple (limit u5) (cursor u0) (data \"\")))"];
+    let expected = ["uint", "(tuple (cursor uint) (data (string-ascii 10)) (limit uint))"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        let type_sig = mem_type_check(good_test).unwrap().0.unwrap();
+        assert_eq!(expected, &type_sig.to_string());
+    }
+}
+
+#[test]
+fn test_string_ascii_as_max_len() {
+    let tests = [
+        "(as-max-len? \"12345\" u5)",
+        "(as-max-len? \"12345\" u8)",
+        "(as-max-len? \"12345\" u4)"];
+    let expected = [
+        "(optional (string-ascii 5))",
+        "(optional (string-ascii 8))",
+        "(optional (string-ascii 4))"];
+
+    for (test, expected) in tests.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&test).unwrap()));
+    }
+}
+
+#[test]
+fn test_string_ascii_concat() {
+    let good = [
+        "(concat \"block\" \"stack\")"];
+    let expected = ["(string-ascii 10)"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&good_test).unwrap()));
+    }
+}
+
+
+#[test]
+fn test_string_utf8_fold() {
+    let good = [
+        "(define-private (get-len (x (string-utf8 1)) (acc uint)) (+ acc u1))
+        (fold get-len u\"blockstack\" u0)",
+        "(define-private (slice (x (string-utf8 1)) (acc (tuple (limit uint) (cursor uint) (data (string-utf8 11)))))
+            (if (< (get cursor acc) (get limit acc))
+                (let ((data (default-to (get data acc) (as-max-len? (concat (get data acc) x) u11))))
+                    (tuple (limit (get limit acc)) (cursor (+ u1 (get cursor acc))) (data data)))
+                acc))
+        (fold slice u\"blockstack\\u{1F926}\" (tuple (limit u5) (cursor u0) (data u\"\")))"];
+    let expected = ["uint", "(tuple (cursor uint) (data (string-utf8 11)) (limit uint))"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        let type_sig = mem_type_check(good_test).unwrap().0.unwrap();
+        assert_eq!(expected, &type_sig.to_string());
+    }
+}
+
+#[test]
+fn test_string_utf8_as_max_len() {
+    let tests = [
+        "(as-max-len? \"1234\\u{1F926}\" u5)",
+        "(as-max-len? \"1234\\u{1F926}\" u8)",
+        "(as-max-len? \"1234\\u{1F926}\" u4)"];
+    let expected = [
+        "(optional (string-ascii 5))",
```

Comment:
Wait, shouldn't this be `(optional (string-utf8 5))`?

Reply:
Hmm -- I think there's something a little more problematic happening here. So utf-8 literals are prefaced with `u"`, so the parser is correctly trying to parse that as an ASCII string. However, the parser is doing some bad character escaping, so it literally parses that the string:

```
1234\u{1F926}
```

Instead of treating the backslash as an escape character.

These tests should be parse errors, something like `UnknownAsciiEscapeCode("\\u")`

Reply:
> Wait, shouldn't this be (optional (string-utf8 5))?

I was using `\"1234\\u{1F926}\"` constructing a `string-ascii` instead of `u\"1234\\u{1F926}\"`, constructing a `string-utf8`. Test fixed with https://github.com/blockstack/stacks-blockchain/pull/1779/commits/0b77404356e87ebba923dcd8468b2184128940bd.

> However, the parser is doing some bad character escaping

This test is confusing because the type-checker can't determine the actual size of a sequence on `as-max-len?`, and always succeeds (could be revisited).
I provided an additional test in https://github.com/blockstack/stacks-blockchain/pull/1779/commits/0b77404356e87ebba923dcd8468b2184128940bd, showing that the behaviour is what we (or at least I) would have been expecting.

Reply:
Yes, but the test revealed a different problem. The string literal parser should not let us use a backslash without the backslash indicating a valid escape sequence.

Reply:
Ha, I might be confused, let's do a step by step.
In our rust code, we're writing:
```rust
"1234\\u{1F926}"
```
because writing directly
```rust 
"1234\u{1F926}"
```
would construct a string "1234🤦🏼‍♂️". As such, we are escaping `\u` with an extra `\`.
Our parser is aware of this issue, and is correcting this with:
https://github.com/blockstack/stacks-blockchain/blob/afc748aa1615ff966675a828f02efd78e092c2d8/src/vm/ast/parser/mod.rs#L296

In the case of ASCII strings, we will end up building the string "1234\u{1F926}" (len 13, because why not), 

in the case of UTF8 strings, we will end up building the string "1234🤦🏼‍♂️" (Len 5).

Reply:
> In the case of ASCII strings, we will end up building the string "1234\u{1F926}" (len 13, because why not),

This, I think, is the problem. While different languages have different approaches to handling "unknown" escape sequences, I think Clarity should be on the pessimistic side of things.

In Javascript:

```
 console.log("a\b\c\d\e\f\n\p\r\l");
```

Outputs:

```
cde

l
```

Where the valid escape characters are parsed correctly, but the invalid ones are just treated as if they were just the characters cde and l.

In Rust:

```
...
    let s = "a\b\c\d\e\f\n\p\r\l";
    println!("{}", s);
...
```

yields a compiler error.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=29eea7b875e90e562aa577ecfc99dbfc

I think we should be making this a parser error.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1779/commits/0e995a99156f85b2b23e918af9af601572b70575